File: 백준/Silver/6374. To the Max/To the Max.py
n = int(input())

dp = [[0 for _ in range(n + 1)] for _ in range(n + 1)]
answer = float('-inf')
i = 0
while 1:
    try:
        line = list(map(int, input().split()))
        for num in line:
            x, y = i % n, i // n
            dp[y + 1][x + 1] = num
            i += 1
    except:
        break


for y in range(1, n + 1):
    temp = 0
    for x in range(1, n + 1):
        temp += dp[y][x]
        dp[y][x] = temp

for x in range(1, n + 1):
    temp = 0
    for y in range(1, n + 1):
        temp += dp[y][x]
        dp[y][x] = temp

for y1 in range(1, n + 1):
    for x1 in range(1, n + 1):
        for y2 in range(0, y1):
            for x2 in range(0, x1):
                size = dp[y1][x1] - dp[y2][x1] - dp[y1][x2] + dp[y2][x2]
                answer = max(answer, size)
print(answer)

    

File: 백준/Silver/6374. To the Max/To the Max.rs
use std::{
    cmp, collections::{ BinaryHeap, HashMap, HashSet, VecDeque }, io::{ stdin, stdout, BufWriter, Read, Write }
};

fn main() {
    let mut s = String::new();

    stdin().read_to_string(&mut s);

    let mut input = s.split_ascii_whitespace();
    let mut output = BufWriter::new(stdout());

    let n: usize = input.next().unwrap().parse().unwrap();
    let mut dp: Vec<Vec<isize>> = vec![vec![0; n + 1]; n + 1];
    for y in 1..n + 1 {
        for x in 1..n + 1 {
            let num: isize = input.next().unwrap().parse().unwrap();
            dp[y][x] = num;
        }
    }

    for y in 0..=n {
        let mut temp = 0;
        for x in 0..=n {
            temp += dp[y][x];
            dp[y][x] = temp;
        }
    }

    for x in 0..=n {
        let mut temp = 0;
        for y in 0..=n {
            temp += dp[y][x];
            dp[y][x] = temp;
        }
    }
    
    let mut answer = isize::MIN;
    for y1 in 1..=n {
        for x1 in 1..=n {
            for y2 in 0..y1 {
                for x2 in 0..x1 {
                    let size = dp[y1][x1] - dp[y2][x1] - dp[y1][x2] + dp[y2][x2];
                    answer = cmp::max(answer, size);
                }
            }
        }
    }
    writeln!(output, "{answer}");
}   
